File: languages/go/entity/identity/datetime_attribute.go
package identity

import (
	"time"

	"github.com/bloock/bloock-sdk-go/v2/internal/bridge/proto"
)

// DatetimeAttribute represents an attribute with a datetime value.
type DatetimeAttribute struct {
	Attribute
}

// NewDatetimeAttribute creates a new DatetimeAttribute instance with the provided key and value.
func NewDatetimeAttribute(key string, value time.Time) DatetimeAttribute {
	a := Attribute{
		Id:    key,
		Value: value.Format(time.RFC3339),
	}
	return DatetimeAttribute{
		Attribute: a,
	}
}

func NewDatetimeAttributeFromProto(s *proto.DateTimeAttribute) DatetimeAttribute {
	if s == nil {
		return DatetimeAttribute{}
	}
	a := Attribute{
		Id:    s.GetId(),
		Value: s.GetValue(),
	}
	return DatetimeAttribute{
		Attribute: a,
	}
}

func (s DatetimeAttribute) ToProto() *proto.DateTimeAttribute {
	return &proto.DateTimeAttribute{
		Id:    s.Attribute.Id,
		Value: s.Attribute.Value.(string),
	}
}

File: languages/js/src/entity/identity/datetime_attribute.ts
import * as identityEntitiesProto from "../../bridge/proto/identity_entities";
import { Attribute } from "./attribute";

/**
 * Represents an attribute with a datetime value.
 */
export class DateTimeAttribute extends Attribute<Date> {
  public toProto(): identityEntitiesProto.DateTimeAttribute {
    return identityEntitiesProto.DateTimeAttribute.fromPartial({
      id: this.id,
      value: this.value.toISOString()
    });
  }

  static fromProto(
    r: identityEntitiesProto.DateTimeAttribute
  ): DateTimeAttribute {
    return new DateTimeAttribute(r.id, new Date(r.value));
  }
}
